vault/external_tests/kv/kv_subkeys_test.go

```diff
@@ -0,0 +1,293 @@
+
+import (
+	"context"
+	"net/http"
+	"testing"
+
+	"github.com/go-test/deep"
+	logicalKv "github.com/hashicorp/vault-plugin-secrets-kv"
+	"github.com/hashicorp/vault/api"
+	vaulthttp "github.com/hashicorp/vault/http"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/vault"
+)
+
+func TestKV_Subkeys_NotFound(t *testing.T) {
+	coreConfig := &vault.CoreConfig{
+		LogicalBackends: map[string]logical.Factory{
+			"kv": logicalKv.VersionedKVFactory,
+		},
+	}
+
+	cluster := vault.NewTestCluster(t, coreConfig, &vault.TestClusterOptions{
+		HandlerFunc: vaulthttp.Handler,
+	})
+
+	cluster.Start()
+	defer cluster.Cleanup()
+
+	cores := cluster.Cores
+
+	core := cores[0].Core
+	c := cluster.Cores[0].Client
+	vault.TestWaitActive(t, core)
+
+	// Mount a KVv2 backend
+	err := c.Sys().Mount("kv", &api.MountInput{
+		Type: "kv-v2",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	req := c.NewRequest("GET", "/v1/kv/subkeys/foo")
+	apiResp, err := c.RawRequestWithContext(context.Background(), req)
+
+	if err == nil || apiResp == nil {
+		t.Fatalf("expected subkeys request to fail, err :%v, resp: %#v", err, apiResp)
+	}
+
+	if apiResp.StatusCode != http.StatusNotFound {
+		t.Fatalf("expected subkeys request to fail with %d status code, resp: %#v", http.StatusNotFound, apiResp)
+	}
+}
+
+func TestKV_Subkeys_Deleted(t *testing.T) {
+	coreConfig := &vault.CoreConfig{
+		LogicalBackends: map[string]logical.Factory{
+			"kv": logicalKv.VersionedKVFactory,
+		},
+	}
+
+	cluster := vault.NewTestCluster(t, coreConfig, &vault.TestClusterOptions{
+		HandlerFunc: vaulthttp.Handler,
+	})
+
+	cluster.Start()
+	defer cluster.Cleanup()
+
+	cores := cluster.Cores
+
+	core := cores[0].Core
+	c := cluster.Cores[0].Client
+	vault.TestWaitActive(t, core)
+
+	// Mount a KVv2 backend
+	err := c.Sys().Mount("kv", &api.MountInput{
+		Type: "kv-v2",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	kvData := map[string]interface{}{
+		"data": map[string]interface{}{
+			"bar": "a",
+		},
+	}
+
+	resp, err := c.Logical().Write("kv/data/foo", kvData)
+	if err != nil {
+		t.Fatalf("write failed, err :%v, resp: %#v\n", err, resp)
+	}
+
+	resp, err = c.Logical().Delete("kv/data/foo")
+	if err != nil {
+		t.Fatalf("delete failed, err :%v, resp: %#v\n", err, resp)
+	}
+
+	req := c.NewRequest("GET", "/v1/kv/subkeys/foo")
+	apiResp, err := c.RawRequestWithContext(context.Background(), req)
+	if resp != nil {
+		defer apiResp.Body.Close()
+	}
+
+	if err == nil || apiResp == nil {
+		t.Fatalf("expected subkeys request to fail, err :%v, resp: %#v", err, apiResp)
+	}
+
+	if apiResp.StatusCode != http.StatusNotFound {
+		t.Fatalf("expected subkeys request to fail with %d status code, resp: %#v", http.StatusNotFound, apiResp)
+	}
+
+	secret, err := api.ParseSecret(apiResp.Body)
+	if err != nil {
+		t.Fatalf("failed to parse resp body, err: %v", err)
+	}
+
+	if subkeys, ok := secret.Data["subkeys"]; !ok || subkeys != nil {
+		t.Fatalf("expected nil subkeys, actual: %#v", subkeys)
```

Comment:
is the !ok check valid for this message? I'm kinda thinking that might be a separate message with a message like `key "subkeys" not found in response`

Reply:
https://github.com/hashicorp/vault/pull/14079/commits/5db99a2940f1a62821de002de02e6cb70aec47c0